numbits.go

```diff
@@ -0,0 +1,45 @@
+package quamina
+
+import (
+	"math"
+)
+
+// numbits is an alternative binary representation of float64 numbers.
+// float64 are stored as (sign | exponent | mantissa)
+// with 1 bit sign, 11 bits exponent, 52 bits mantissa
+// They can be represented as [8]byte or as string and can be created from
+// these representations.
+// All possible float64 values are representable as numbits.
+// numbits were implemented by Arne Hormann for Quamina; he later discovered
+// that an equivalent representation was used long ago in the disk format of DB2.
+//
+// Arne's implementation carefully handled NaN, -0, and infinities, but
+// Quamina ignores those issues because a combination of JSON rules and
+// Quamina's parsers prevent those values from occurring.
+type numbits uint64
+
+// numbitsFromFloat64 converts a float64 value to its numbits representation.
+func numbitsFromFloat64(f float64) numbits {
+	u := math.Float64bits(f)
+	// transform without branching (inverse of numbits.Float64):
+	// if high bit is 0, xor with sign bit 1 << 63, else negate (xor with ^0)
+	mask := (u>>63)*^uint64(0) | (1 << 63)
+	return numbits(u ^ mask)
+}
+
+// toUTF8 turns a numbits into 10 bytes of UTF-8 encoded via Base-256
+// code copied with thanks from a sample by Axel Wagner
+func (nb numbits) toUTF8() []byte {
+	var b [10]byte
```

Comment:
If performance critical, this could be unrolled as shown in https://github.com/timbray/quamina/issues/334#issuecomment-2295324890 - both versions do the same big endian base128 encoding.
```
be128 := []byte {
  byte(num >> 57) & 0x7f, // 57 bits remaining
  byte(num >> 50) & 0x7f, // 50 bits remaining
  byte(num >> 43) & 0x7f, // 43 bits remaining
  byte(num >> 36) & 0x7f, // 36 bits remaining
  byte(num >> 29) & 0x7f, // 29 bits remaining
  byte(num >> 22) & 0x7f, // 22 bits remaining
  byte(num >> 15) & 0x7f, // 15 bits remaining
  byte(num >> 8) & 0x7f, //  8 bits remaining
  byte(num >> 1) & 0x7f, //  1 bit  remaining
  byte(num & 1) << 6, // last bit (high bit of lower 7 bits)
}
```
and "math/bits" can be used to slice it: `(bits.TrailingZeros64(uint64(nb)) + 6) / 7` can determine the end without looking.

Reply:
I prefer the loop construct, I think it's roughly as readable and takes less screen space.  Or am I missing something?

As for `TrailingZeros64`, I think there's a problem, which is the case when the numbits reduces to just `[0]` (the smallest possible number I guess).  Because removing all the trailing zeroes would reduce this to a n empty slice and the automaton needs at least one byte to match on. That's why the current loop has `k>0` not `>=0`. But, I should comment that because it's non-obvious.

Reply:
numbit 0 is NaN, cannot happen in JSON ☺️

Reply:
Heh, OK. But I better say that in a comment…

Reply:
also, base 256 would be a plain byte. it's big endian base 128 (only the lower 7 bits are used for all bytes, byte order is big endian).

Reply:
Oh you're right.  Off by 1.

Reply:
Anyhow, I tried out `bits.TrailingZeros64` but the code was a bit awkward and the +6 / 7 nonintuitive, but I did go back and optimize a bit; first count back after 0-byte encodings, and leave everything set up for a separate loop that allocates just the right size and copies them in.

Reply:
it still says Base-256 in the comment

Reply:
I also made a mistake, there. the +6/7 is only needed if we go from left to right. 
It should have been `be128[:len(be128)-bits.TrailingZeros64(uint64(nb))]`